src/if_statements.py

```diff
@@ -570,6 +602,213 @@ def try_make_if_condition(
     return (cond, if_node, else_node)
 
 
+def try_build_implicit_switch(
+    context: Context, start: Node, end: Node
+) -> Optional[SwitchStatement]:
+    """
+    Look for implicit switch-like structures from nested ConditionalNodes & SwitchNodes.
+    If one is found, return the corresponding SwitchStatement; otherwise, return None.
+
+    Both IDO & GCC can convert switch statements into a tree of comparisons & jump tables.
+    We try to identify the largest contiguous tree of ConditionalNodes & SwitchNodes that
+    all compare the same variable (`var_expr`). SwitchNodes and `(x == N)` ConditionalNodes
+    represent jumps to specific case labels, whereas comparisons like `(x < N)` are
+    primarily used to manage the overall tree depth.
+    """
+    # The start node must either be an if or a switch
+    if not isinstance(start, (ConditionalNode, SwitchNode)):
+        return None
+
+    assert end in start.postdominators
+    start_block_info = get_block_info(start)
+    control_expr = switch_guard_expr(start)
+    var_expr: Optional[Expression] = None
+
+    # Try to extract the switch's control expression
+    if control_expr is not None:
+        # `if (x >= len(jump_table))`: var_expr is `x`
+        var_expr = control_expr
+    elif start_block_info.switch_control is not None:
+        # `switch(x)`: var_expr is `x`
+        var_expr = start_block_info.switch_control.control_expr
+    elif start_block_info.branch_condition is not None:
+        start_cond = simplify_condition(start_block_info.branch_condition)
+        if (
+            start_cond is not None
+            and isinstance(start_cond, BinaryOp)
+            and isinstance(start_cond.right, Literal)
+        ):
+            # `if (x == N)`: var_expr is `x`
+            # The `start_cond.op` is checked in the first iter of the while loop below
+            var_expr = start_cond.left
+    if var_expr is None:
+        return None
+
+    # Unwrap EvalOnceExpr's and Cast's; ops like `<=` always include an `(s32)` cast
+    uw_var_expr = early_unwrap_ints(var_expr)
+    # Nodes we need to visit, initially just the `start` node
+    node_queue: List[Node] = [start]
+    # Nodes we have already visited, to avoid infinite loops
+    visited_nodes: Set[Node] = set()
+    # Nodes that have no statements, and should be marked as emitted if we emit a SwitchStatement.
+    # A ConditionalNode like `if (x == N)` isn't directly emitted; it's replaced by a `case N:` label
+    # This also includes empty BasicNodes & ReturnNodes that jump directly to the end
+    nodes_to_mark_emitted: Set[Node] = set()
+    # Map of label -> node. Similar to SwitchNode.cases, but the labels may not be contiguous
+    cases: Dict[int, Node] = {}
+    # The `default:`-labeled node, if found
+    default_node: Optional[Node] = None
+    # Require at least one ConditionalNode that is not a guard for a SwitchNode
+    has_non_guard_conditional = False
```

Comment:
Can't we have a switch that's split into two jump tables and no other conditional nodes?

Reply:
Hmm, I don't know if the compiler could/would do that? But you might be right...

The intent of this check was to avoid handling the case of just a single SwitchNode & it's ConditionalNode guard, because detecting the default node in this function was less accurate than the existing logic. 

I'm going to work on this some more to see if I can make the default node matching better. It would be good to get that working, so that the logic in `build_flowgraph_between` (`elif switch_guard_expr(curr_start) is not None: ...`) could be removed in the future?

Reply:
Apparently IDO doesn't:
```c
void foo(void) {
    switch (x) {
    case 0:
    case 1:
    case 2:
    case 3:
    case 4:
    case 100:
    case 101:
    case 102:
    case 103:
    case 104:
        bar();
    }
}
```
results in
```
00000000 <foo>:
   0:	27bdffd8 	addiu	sp,sp,-40
   4:	afb00018 	sw	s0,24(sp)
   8:	3c100000 	lui	s0,0x0
			8: R_MIPS_HI16	x
   c:	8e100000 	lw	s0,0(s0)
			c: R_MIPS_LO16	x
  10:	afbf001c 	sw	ra,28(sp)
  14:	2a010005 	slti	at,s0,5
  18:	1420000b 	bnez	at,48 <foo+0x48>
  1c:	00000000 	nop
  20:	260eff9c 	addiu	t6,s0,-100
  24:	2dc10005 	sltiu	at,t6,5
  28:	10200012 	beqz	at,74 <foo+0x74>
  2c:	00000000 	nop
  30:	000e7080 	sll	t6,t6,0x2
  34:	3c010000 	lui	at,0x0
			34: R_MIPS_HI16	.rodata
  38:	002e0821 	addu	at,at,t6
  3c:	8c2e0000 	lw	t6,0(at)
			3c: R_MIPS_LO16	.rodata
  40:	01c00008 	jr	t6
  44:	00000000 	nop
  48:	2e010005 	sltiu	at,s0,5
  4c:	10200009 	beqz	at,74 <foo+0x74>
  50:	00000000 	nop
  54:	00107880 	sll	t7,s0,0x2
  58:	3c010000 	lui	at,0x0
			58: R_MIPS_HI16	.rodata
  5c:	002f0821 	addu	at,at,t7
  60:	8c2f0014 	lw	t7,20(at)
			60: R_MIPS_LO16	.rodata
  64:	01e00008 	jr	t7
  68:	00000000 	nop
  6c:	0c000000 	jal	0 <foo>
			6c: R_MIPS_26	bar
  70:	00000000 	nop
  74:	8fbf001c 	lw	ra,28(sp)
  78:	8fb00018 	lw	s0,24(sp)
  7c:	27bd0028 	addiu	sp,sp,40
  80:	03e00008 	jr	ra
  84:	00000000 	nop
```
i.e.
```
if (x < 5) {
  if (x < 5) jtbl;
} else {
  if (x - 100 < 5) jtbl;
}
```
which is a bit silly. GCC on the other hand doesn't believe in multiple jump tables: https://decomp.me/scratch/dJz3D

Reply:
(Just to be clear, the updated code with the `irregular_comparison_count` variable is more accepting of these types of things, and would accept a pair of jump tables with no other nodes. It doesn't seem to come up in practice, but it doesn't hurt)